tsinfer/inference.py

```diff
@@ -584,6 +549,228 @@ def match_ancestors(
     return ts
 
 
+def match_ancestors_batch_init(
+    working_dir,
+    sample_data_path,
+    ancestor_data_path,
+    min_work_per_job,
+    *,
+    sgkit_samples_mask_name=None,
+    sites_mask_name=None,
+    recombination_rate=None,
+    mismatch_ratio=None,
+    path_compression=True,
+    # Deliberately undocumented parameters below
+    recombination=None,  # See :class:`Matcher`
+    mismatch=None,  # See :class:`Matcher`
+    precision=None,
+    engine=constants.C_ENGINE,
+    extended_checks=False,
+    time_units=None,
+    record_provenance=True,
+):
+    # Create work dir
+    pathlib.Path(working_dir).mkdir(parents=True, exist_ok=True)
+    ancestors = formats.AncestorData.load(ancestor_data_path)
+    sample_data = formats.SgkitSampleData(
+        sample_data_path,
+        sgkit_samples_mask_name=sgkit_samples_mask_name,
+        sites_mask_name=sites_mask_name,
+    )
+    ancestors._check_finalised()
+    sample_data._check_finalised()
+
+    matcher = AncestorMatcher(
+        sample_data,
+        ancestors,
+    )
+    ancestor_grouping = []
+    ancestor_lengths = ancestors.ancestors_length
+    for group_index, group_ancestors in matcher.group_by_linesweep().items():
+        # Make ancestor_ids JSON serialisable
+        group_ancestors = list(map(int, group_ancestors))
+        partitions = []
+        current_partition = []
+        current_partition_work = 0
+        # TODO: Can do better here by packing ancestors
+        # into as equal sized partitions as possible
+        if group_index == 0:
+            partitions.append(group_ancestors)
+        else:
+            total_work = sum(ancestor_lengths[ancestor] for ancestor in group_ancestors)
+            min_work_per_job_group = min_work_per_job
+            if total_work / 1000 > min_work_per_job:
+                min_work_per_job_group = total_work / 1000
+            for ancestor in group_ancestors:
+                if (
+                    current_partition_work + ancestor_lengths[ancestor]
+                    > min_work_per_job_group
+                ):
+                    partitions.append(current_partition)
+                    current_partition = [ancestor]
+                    current_partition_work = ancestor_lengths[ancestor]
+                else:
+                    current_partition.append(ancestor)
+                    current_partition_work += ancestor_lengths[ancestor]
+            partitions.append(current_partition)
+        # Make directories for the path data
+        if len(partitions) > 1:
+            os.mkdir(os.path.join(working_dir, f"group_{group_index}"))
+        group = {
+            "ancestors": group_ancestors,
+            "partitions": partitions if len(partitions) > 1 else None,
+        }
+        ancestor_grouping.append(group)
+
```

Comment:
We should implement this as a dataclass at some point, as this will be fragile across version changes etc. Dict is fine for now, but let's add a comment for future references.